aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -35,18 +91,155 @@ struct _cpu_scatter_gather_dim_loop {
       );
     }
   }
+
+  template <typename scalar_t, typename func_t>
+  void operator()(
+    scalar_t* self_data, int64_t self_dim_stride,
+    int64_t* index_data, int64_t index_dim_stride,
+    Scalar value,
+    int64_t dim, int64_t index_dim_size,
+    int64_t index_upper_bound,
+    func_t f
+  ) {
+
+    for (int64_t i = 0; i < index_dim_size; ++i) {
+      int64_t idx_dim = index_data[i * index_dim_stride];
+      // we are not putting idx_dim in the error message because it disables
+      // loop optimization in clang-7
+      TORCH_CHECK(idx_dim >= 0 && idx_dim < index_upper_bound,
+        "index ", index_data[i * index_dim_stride],
+        " is out of bounds for dimension ", dim,
+        " with size ", index_upper_bound
+      );
+      auto temp = value.to<scalar_t>();
+      f(
+        self_data + (is_scatter_like ? idx_dim : i) * self_dim_stride, &temp
+      );
+    }
+  }
 };
 
+
 template <bool is_scatter_like = true>
 struct cpu_scatter_gather_base_kernel {
   template <typename func_t>
-  void operator()(
-    Tensor& self, int64_t dim,
+  void operator()(Tensor& self, int64_t dim,
+    const Tensor& index, Scalar& value,
+    const std::string& method_name,
+    bool serial_exec, func_t kernel_func) {
+    // no-op if index is empty
+    if (index.numel() == 0) {
+      return;
+    }
+
+    dim = maybe_wrap_dim(dim, self.dim());
+
+    if (is_scatter_like) {
+      scatter_shape_check(self, dim, index, self);
+    }
+    else {
+      gather_shape_check(self, dim, index, self);
+    }
+
+    auto index_sizes = ensure_nonempty_vec(index.sizes().vec());
+    auto index_strides = ensure_nonempty_vec(index.strides().vec());
+
+    // `dim` is traversed in the kernel,
+    // that is why index.stride(dim) = 0 and index.size(dim) = 1.
+    // Also, index.size(dim) = 1 makes sure that TensorIterator.DimCounter
+    // has the following form : (i_1,..., i_{dim-1}, 0, i_{dim+1},...,i_n).
+    index_sizes[dim] = 1;
+    index_strides[dim] = 0;
+
+    // set self.shape = src.shape = index.shape,
+    // this defines the number of elements to iterate over,
+    // and set self.stride(dim) = src.stride(dim) = 0,
+    // because `dim` is traversed in the kernel.
+    auto self_restrided = restride_dim(self, dim, index_sizes);
+    auto index_restrided = index.as_strided(index_sizes, index_strides);
```

Comment:
rebase, this chunk of code is not needed now that TensorIterator supports squashed dimensions.

Reply:
Serial_exec is also about to get merged.

Reply:
Also, serial_exec removal is about to get merged.

Reply:
The code failed when I removed this and replaced `self_restrided` with `self`.